cmd/juju/application/bundle.go

```diff
@@ -881,3 +887,90 @@ func isErrRelationExists(err error) bool {
 	// the string when a specific cause is available.
 	return params.IsCodeAlreadyExists(err) || strings.HasSuffix(err.Error(), "relation already exists")
 }
+
+func processBundleIncludes(baseDir string, data *charm.BundleData) error {
+
+	for app, appData := range data.Applications {
+		for key, value := range appData.Options {
+			result, processed, err := processValue(baseDir, value)
+			if err != nil {
+				return errors.Annotatef(err, "processing options value %s for application %s", key, app)
+			}
+			if processed {
+				appData.Options[key] = result
+			}
+		}
+		for key, value := range appData.Annotations {
+			result, processed, err := processValue(baseDir, value)
+			if err != nil {
+				return errors.Annotatef(err, "processing annotation value %s for application %s", key, app)
+			}
+			if processed {
+				appData.Annotations[key] = result.(string)
+			}
+		}
+	}
+
+	for machine, machineData := range data.Machines {
+		for key, value := range machineData.Annotations {
+			result, processed, err := processValue(baseDir, value)
+			if err != nil {
+				return errors.Annotatef(err, "processing annotation value %s for machine %s", key, machine)
+			}
+			if processed {
+				machineData.Annotations[key] = result.(string)
+			}
+		}
+	}
+	return nil
+}
+
+func processValue(baseDir string, v interface{}) (interface{}, bool, error) {
+
+	const (
+		includeFile   = "include-file://"
+		includeBase64 = "include-base64://"
+	)
+
+	value, ok := v.(string)
+	if !ok {
+		// Not a string, just return it unchanged.
+		return v, false, nil
+	}
+
+	encode := false
+	readFile := false
+	filename := ""
+
+	if strings.HasPrefix(value, includeFile) {
+		readFile = true
+		filename = value[len(includeFile):]
+	} else if strings.HasPrefix(value, includeBase64) {
+		encode = true
+		readFile = true
+		filename = value[len(includeBase64):]
+	}
+
+	if !readFile {
+		// Unchanged, just return it.
+		return v, false, nil
+	}
+
+	if !filepath.IsAbs(filename) {
+		filename = filepath.Clean(filepath.Join(baseDir, filename))
```

Comment:
This value is called `baseDir` here, but it's being passed in as `bundleFilePath` from `deployBundle` - will it be the path of the bundle file, rather than the containing directory?